packit/utils/obs_helper.py

```diff
@@ -0,0 +1,454 @@
+    person_elem.attrib["role"] = "maintainer"
+
+    for elem in (title_elem, descr_elem, person_elem):
+        root.append(elem)
+
+    for repo in repos:
+        (repo_elem := ET.Element("repository")).attrib["name"] = repo.name
+
+        for path_entry in repo.path:
+            (path_elem := ET.Element("path")).attrib["project"] = path_entry.project
+            path_elem.attrib["repository"] = path_entry.repository
+            repo_elem.append(path_elem)
+
+        for arch in repo.arch:
+            (arch_elem := ET.Element("arch")).text = arch
+            repo_elem.append(arch_elem)
+
+        root.append(repo_elem)
+
+    return root
+
+
+def parse_changelog_subject(line: str):
+    date: list[str]
+    author: list[str] = []
+    email: str = ""
+    version: str = ""
+
+    toks = [tok.strip() for tok in line.split(" ") if tok.strip()]
+
+    toks.pop(0)  # Remove leading "*"
+    date = toks[:4]
+
+    del toks[:4]
+    for index, element in enumerate(toks):
+        if not element.startswith("<"):
+            author.append(element)
+        else:
+            email += element
+            del toks[: index + 1]
+            break
+
+    if toks and toks[0] == "-":
+        toks.pop(0)
+
+    if toks:
+        version = toks[0]
+
+    return date, author, email, version
+
+
+def format_changelog(changelog_str: str) -> str:
+    """Converts spec changelog to OBS .changes format"""
+    output: list[str] = []
+    changelog_time = "12:00:00 UTC"
+    changelog = changelog_str.splitlines()
+
+    for line in changelog:
+        if not line.strip():
+            continue
+        if line.strip().startswith("*"):
+            if output:
+                output[-1] += "\n"
+
+            date, author, email, version = parse_changelog_subject(line)
+            date.insert(3, changelog_time)
+
+            output.append("-" * 68)
+            output.append(f"{' '.join(date)} - {' '.join(author)} {email}\n")
+            if version:
+                output.append(f"- {version}")
+        else:
+            if re.match(r"-\s[0-9]+[\.\-]*", line):  # Match version line
+                output.append(line)
+            else:
+                output.append(f"  {line}")
+
+    return "\n".join(output)
+
+
+def create_changes_file(package_dir: Path) -> None:
+    """
+    Creates a changelog file from a by copying the changelog section from the spec file.
+    Sets release to 0 since obs handles release numbers
+
+    Args:
+        package_dir (Path): Path to the directory containing the package spec file.
+
+    Returns:
+        None
+    """
+    specfile: str
+    content: str
+    release_line = "Release:    0"
+
+    for file in list(filter(os.path.isfile, os.listdir(package_dir))):
+        if file.endswith(".spec"):
+            specfile = file
+            break
+
+        if file.endswith(".changes"):
+            logger.info(".changes file already exists in package")
+            return
```

Comment:
How are you guaranteed that you will find the changelog before finding the specfile?

Reply:
It's just a convenient way of checking if a `.changes` file already exist. At this point the spec file is guaranteed to exist.

Reply:
I know what you're doing, it's just not correct :)

From the [docs](https://docs.python.org/3/library/os.html#os.listdir):
> Return a list containing the names of the entries in the directory given by _path_. The list is in **arbitrary** order,

Let's assume that I have a package `hello-world`, and I get out of the `list(filter(…))`:
```
hello-world.spec
hello-world.changes
```

I find the specfile, I **break** out of the loop, and few lines below I'll **rewrite** the `hello-world.changes`.

OH! There is a return with `Project has no changelog` :facepalm: I _guess_ it's fine then…